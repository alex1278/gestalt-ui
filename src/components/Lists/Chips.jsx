import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Error, Caption } from 'components/Typography';
import TextField from 'components/Fields/TextField';
import { insertItem, removeItem } from 'util/helpers/lists';
import Chip from './Chip';

const List = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

class Chips extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    meta: PropTypes.object.isRequired,
    prefix: PropTypes.string,
    input: PropTypes.object,
    ignorePrefixValidation: PropTypes.bool,
    label: PropTypes.string,
    forceLowerCase: PropTypes.bool,
    helpText: PropTypes.string,
  };

  static defaultProps = {
    input: {},
    prefix: null,
    ignorePrefixValidation: false,
    label: 'Entry',
    forceLowerCase: false,
    helpText: null,
  };

  state = { item: '', touched: false };

  addItem = () => {
    let prefixedItem;

    if (this.props.prefix) {
      prefixedItem = `${this.props.prefix}.${this.state.item}`;
    } else {
      prefixedItem = this.state.item;
    }

    const isDupe = this.props.input.value && this.props.input.value.find(item => item === prefixedItem);

    if (this.state.item && !isDupe) {
      this.props.input.onChange(insertItem(this.props.input.value, prefixedItem));
      this.setState({ item: '' });
    }
  }

  removeItem = (value) => {
    this.props.input.onChange(removeItem(this.props.input.value, value));
  }

  handleChange = (e) => {
    if (this.props.forceLowerCase) {
      this.setState({ item: e.target.value.toLowerCase().trim() });
    } else {
      this.setState({ item: e.target.value.trim() });
    }
  }

  handleTouched = () => {
    this.setState({ touched: true });
  }

  handleEnter = (e) => {
    if (e.key === 'Enter') {
      e.preventDefault();
      this.addItem();
    }
  }

  render() {
    const { id, label, input, meta: { error }, helpText } = this.props;

    return (
      <React.Fragment>
        <TextField
          id={`${id}-list-table`}
          placeholder={label}
          type="text"
          fullWidth
          onChange={this.handleChange}
          value={this.state.item}
          disabled={!this.props.prefix && !this.props.ignorePrefixValidation}
          onFocus={this.handleTouched}
          onBlur={this.addItem}
          onKeyDown={this.handleEnter}
          helpText="press ENTER to add a new entry"
        />
        {helpText && <Caption light>{helpText}</Caption>}
        <List maxHeight="184px">
          {Array.isArray(input.value) && input.value.map((item, i) => <Chip key={`${item}--${i}`} item={item} onRemove={this.removeItem} />)}
        </List>
        {(this.state.touched && !!error) && <Error>{error}</Error>}
      </React.Fragment>
    );
  }
}

export default Chips;
