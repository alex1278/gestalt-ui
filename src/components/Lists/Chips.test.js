import React from 'react';
import { render } from 'react-testing-library';
import Chips from './Chips';

describe('Chips', () => {
  it('mounts with basic props', () => {
    const { container } = render(<Chips id="test" meta={{ error: null, touched: false }} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
