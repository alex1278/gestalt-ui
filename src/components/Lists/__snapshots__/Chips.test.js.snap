// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips mounts with basic props 1`] = `
<div
  class="MuiFormControl-root-1 MuiFormControl-marginDense-3 MuiFormControl-fullWidth-4 "
>
  <div
    class="MuiInputBase-root-18 MuiOutlinedInput-root-5 MuiInputBase-disabled-21 MuiOutlinedInput-disabled-7 MuiInputBase-fullWidth-27 MuiInputBase-formControl-19 MuiInputBase-marginDense-25"
  >
    <fieldset
      aria-hidden="true"
      class="MuiPrivateNotchedOutline-root-35 MuiOutlinedInput-notchedOutline-12"
      style="padding-left: 8px;"
    >
      <legend
        class="MuiPrivateNotchedOutline-legend-36"
        style="width: 0.01px;"
      >
        <span>
          â€‹
        </span>
      </legend>
    </fieldset>
    <input
      aria-describedby="test-list-table-helper-text"
      aria-invalid="false"
      class="MuiInputBase-input-28 MuiOutlinedInput-input-13 MuiInputBase-disabled-21 MuiOutlinedInput-disabled-7 MuiInputBase-inputMarginDense-29 MuiOutlinedInput-inputMarginDense-14"
      disabled=""
      id="test-list-table"
      placeholder="Entry"
      type="text"
      value=""
    />
  </div>
  <p
    class="MuiFormHelperText-root-37 MuiFormHelperText-contained-41 MuiFormHelperText-marginDense-40 MuiFormHelperText-disabled-39"
    id="test-list-table-helper-text"
  >
    press ENTER to add a new entry
  </p>
</div>
`;
